arch-design-questions.md

```diff
@@ -0,0 +1,138 @@
+    read from session until all message exhausted
+    close session
+```
+
+**get next session** can be handled by simply allowing the [single use client](#single-use-clients)
+to take a sentinel value for the session ID, allowing the server to just hand us the next unlocked session.
+
+We anticipate that this will be a fairly common pattern. To simplify the experience for the user each
+language will have some form of a **session manager**.
+
+This client is a higher-level client that internally manages the connection on behalf of the user, does
+manage state and is distinct from the [single use clients](single-use-clients).
+
+Existing solutions:
+
+- JS: [`SessionManager` (has not been reviewed for production-readiness)](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/servicebus/service-bus/src/session/sessionManager.ts)
+- .NET: [`ServiceBusProcessor`](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Processor/ServiceBusProcessorClient.cs)
+- Java/Python - no solution currently.
+
+# Connection sharing via top-level object
+
+Connection sharing is enabled by _default_ in our current design by virtue of a 
+top level `ServiceBusClient` object.
+
+The `ServiceBusClient` object is created with either connection string
+or TokenCredential information but is _not_ entity specific.
+
+So typical usage will be somewhat like this:
+
+```csharp
+class ServiceBusClient {
+  ServiceBusClient(connectionstring or TokenCredential/host)
+
+  // any receiver/sender instances created through here will share the
+  // same AMQP connection which is owned by the ServiceBusClient
+  ServiceBusReceiver GetReceiver(queuename);
+  ServiceBusReceiver GetReceiver(topicname, subscriptionname);  
+  // optional-session-id, if null, means 'get next available session'
+  // also, actually locks the session at this point.
+  Task<ServiceBusReceiver> GetSessionReceiver(queuename, optional-sessionid);
+  ServiceBusSender GetSender(queuename or topic name);
+}
+```
+
+and the user experience is:
+
+```csharp
+ServiceBusClient client = new ServiceBusClient(connectionstring);
+
+var queueReceiver = client.GetReceiver("queue-name");
+
+// shares same link as `queueReceiver`
+var topicSender = client.GetSender("topic-name");
+
+// closes the receiver link (but not connection)
+await queueReceiver.CloseAsync();
+// closes sender link (but not connection)
+await topicSender.CloseAsync();
+
+// closes the AMQP connection
+await client.CloseAsync();
+```
+
+# Different messages for sending vs receiving
+
+Messages will have separate classes/interfaces for messages that are sent (`Message`)
+vs messages that are received (`ReceivedMessage`). 
+
+This allows us to avoid including fields or actions (like settling, that only applies 
+to received messages).
+
+Messages returned from the peek() API should not contain methods for 
+message settlement. (ie, complete(), abandon()).
+
+TBD: do we make this a:
```

Comment:
Need to resolve this. We can do separate types easily in JS so I vote for "separate types" and making this a compile time thing.

Reply:
.NET doesn't have service methods on the message objects. In order to make it a compile time thing, we could still introduce a new type between the most derived ReceivedMessage and the least derived Message. I think it is not worth doing as Peek isn't a common operation, and we would be adding a new type for just this one operation. I would prefer to stick with runtime validation that we already have for settlement. In fact, we already throw in settlement methods if the ReceiveMode was ReceiveAndDelete, so we would just be adding one more constraint. @ShivangiReja @jsquire thoughts on this?

Reply:
Since we can not do message settlement on messages from peek() API . Either Receiver have to throw runtime exception is someone called abandon() or we can create `SBPeekMessage` as new type. 

Separate Type will be better user experience as well.

Reply:
@JoshLove-msft  - updated to include your probable behavior of throwing a runtime exception (to avoid introducing additional types, right?)

Reply:
@hemanttanwar - which way do you lean on this? I assume you and .NET will want to line up.

Reply:
I'm with Josh on this.

Reply:
@richardpark-msft  I would be okay with throwing Exception , but need to get reviewed with Jonathan.